File: src/main/java/com/marsraver/WledFx/animation/SnakeAnimation.java
package com.marsraver.WledFx.animation;

/**
 * Snake animation that travels across the grid
 */
public class SnakeAnimation implements LedAnimation {
    
    private int combinedWidth;
    private int combinedHeight;
    private int snakePosition;
    
    @Override
    public void init(int combinedWidth, int combinedHeight) {
        this.combinedWidth = combinedWidth;
        this.combinedHeight = combinedHeight;
        this.snakePosition = 0;
    }
    
    @Override
    public boolean update(long now) {
        // Move snake forward
        snakePosition = (snakePosition + 1) % (combinedWidth * combinedHeight);
        return true; // Keep running
    }
    
    @Override
    public int[] getPixelColor(int x, int y) {
        int ledIndex = y * combinedWidth + x;
        int distance = Math.abs(ledIndex - snakePosition);
        
        // Create snake effect with tail
        if (distance == 0) {
            // Head - bright green
            return new int[]{0, 255, 0};
        } else if (distance <= 5) {
            // Tail - fade to dark green
            int intensity = 255 - (distance * 40);
            intensity = Math.max(0, Math.min(255, intensity));
            return new int[]{0, intensity, 0};
        } else {
            // Off
            return new int[]{0, 0, 0};
        }
    }
    
    @Override
    public String getName() {
        return "Snake";
    }
}

File: src/main/kotlin/com/marsraver/wledfx/animation/SnakeAnimation.kt
package com.marsraver.wledfx.animation

import com.marsraver.wledfx.color.RgbColor
import com.marsraver.wledfx.color.ColorUtils
import kotlin.math.abs

/**
 * Snake animation
 */
class SnakeAnimation : BaseAnimation() {

    private var snakePosition: Int = 0

    override fun getName(): String = "Snake"
    override fun is1D(): Boolean = true
    override fun is2D(): Boolean = false

    override fun onInit() {
        snakePosition = 0
    }

    override fun update(now: Long): Boolean {
        snakePosition = (snakePosition + 1) % pixelCount
        return true
    }

    override fun getPixelColor(x: Int, y: Int): RgbColor {
        val ledIndex = getPixelIndex(x, y)
        val distance = abs(ledIndex - snakePosition)

        return when {
            distance == 0 -> paramColor
            distance <= 5 -> {
                val intensity = (255 - distance * 40).coerceIn(0, 255)
                val factor = intensity / 255.0
                ColorUtils.scaleBrightness(paramColor, factor)
            }
            else -> RgbColor.BLACK
        }
    }
}
